### PR TITLE
{AKS} `az aks create/az aks nodepool add`: Update the default value of option `--max-pods` and `--node-osdisk-size` to `None`

### PR DESCRIPTION
**Pull request by FumingZhang**
**Related command**

- az aks create
- az aks nodepool add

**Description**

Update the default value of option `--max-pods` and `--node-osdisk-size` to `None`. As the type of these two options is int, their default values were set to 0 previously, but this does not match the definition of the corresponding attribute.

**Testing Guide**

**History Notes**

[Component Name 1] BREAKING CHANGE: `az command a`: Make some customer-facing breaking change
[Component Name 2] `az command b`: Add some customer-facing feature

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [ ] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [ ] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [ ] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
AKS

File: src/azure-cli/azure/cli/command_modules/acs/tests/latest/test_agentpool_decorator.py [hunk 1]
```diff
@@ -1065,7 +1065,7 @@ def common_get_max_pods(self):
             DecoratorMode.CREATE,
             self.agentpool_decorator_mode,
         )
-        self.assertEqual(ctx_1.get_max_pods(), None)
+        self.assertEqual(ctx_1.get_max_pods(), 0)
```
Comment (yanzhudd) [inline src/azure-cli/azure/cli/command_modules/acs/tests/latest/test_agentpool_decorator.py:1068]: Please test the output of the function if inputting `max_pods` with None value.

**Comment by yanzhudd** [comment]
Please add test for `az aks create/update` with not inputting `--max-pods/--node-osdisk-size` to check if the response meets your expectation.

File: src/azure-cli/azure/cli/command_modules/acs/custom.py [hunk 2]
```diff
@@ -2170,11 +2170,11 @@ def aks_agentpool_add(
     tags=None,
     node_taints=None,
     node_osdisk_type=None,
-    node_osdisk_size=0,
+    node_osdisk_size=None,
     max_surge=None,
     mode=CONST_NODEPOOL_MODE_USER,
     scale_down_mode=CONST_SCALE_DOWN_MODE_DELETE,
-    max_pods=0,
+    max_pods=None,
```
Comment (yanzhudd) [inline src/azure-cli/azure/cli/command_modules/acs/custom.py:2177]: Let me check if this PR will not only change the default value in `az aks create/update` command, but also other command like `az aks nodepool add`?
If so, please add this part of info to the title and description of this PR.

Comment (FumingZhang) [reply src/azure-cli/azure/cli/command_modules/acs/custom.py:2177]: Good catch, the change is applied to command `az aks create` and `az aks nodepool add`

Comment (FumingZhang) [reply src/azure-cli/azure/cli/command_modules/acs/custom.py:2177]: updated PR title

File: src/azure-cli/azure/cli/command_modules/acs/tests/latest/test_aks_commands.py [hunk 3]
```diff
@@ -147,6 +147,8 @@ def test_aks_create_default(self, resource_group, resource_group_location):
             self.check('agentPoolProfiles[0].nodeLabels.label2', 'value2'),
             self.check('agentPoolProfiles[0].tags.tag1', 'tv1'),
             self.check('agentPoolProfiles[0].tags.tag2', 'tv2'),
+            self.check('agentPoolProfiles[0].maxPods', 110),
```
Comment (FumingZhang) [inline src/azure-cli/azure/cli/command_modules/acs/tests/latest/test_aks_commands.py:150]: Queued [live test pipeline](https://dev.azure.com/msazure/CloudNativeCompute/_build/results?buildId=78530549&view=results) to validate the change.

File: src/azure-cli/azure/cli/command_modules/acs/tests/latest/test_aks_commands.py [hunk 4]
```diff
@@ -591,7 +591,8 @@ def test_aks_create_default_service(self, resource_group, resource_group_locatio
             self.exists('nodeResourceGroup'),
             self.check('provisioningState', 'Succeeded'),
             self.check('tags.key1', 'value1'),
-            self.check('servicePrincipalProfile.clientId', sp_name)
+            self.check('servicePrincipalProfile.clientId', sp_name),
+            self.check('agentPoolProfiles[0].maxPods', 100),
```
Comment (FumingZhang) [inline src/azure-cli/azure/cli/command_modules/acs/tests/latest/test_aks_commands.py:595]: Queued [live test](https://dev.azure.com/msazure/CloudNativeCompute/_build/results?buildId=78530664&view=results) to validate the change.

File: src/azure-cli/azure/cli/command_modules/acs/tests/latest/test_aks_commands.py [hunk 5]
```diff
@@ -2721,6 +2722,7 @@ def test_aks_create_with_ephemeral_disk(self, resource_group, resource_group_loc
         self.cmd(create_cmd, checks=[
             self.check('provisioningState', 'Succeeded'),
             self.check('agentPoolProfiles[0].osDiskType', 'Ephemeral'),
+            self.check('agentPoolProfiles[0].osDiskSizeGb', 60),
```
Comment (FumingZhang) [inline src/azure-cli/azure/cli/command_modules/acs/tests/latest/test_aks_commands.py:2725]: Queued [live test](https://dev.azure.com/msazure/CloudNativeCompute/_build/results?buildId=78530664&view=results) to validate the change.

File: src/azure-cli/azure/cli/command_modules/acs/tests/latest/test_aks_commands.py (hunk 3 above)
Comment (FumingZhang) [reply src/azure-cli/azure/cli/command_modules/acs/tests/latest/test_aks_commands.py:150]: test passed

File: src/azure-cli/azure/cli/command_modules/acs/tests/latest/test_aks_commands.py (hunk 5 above)
Comment (FumingZhang) [reply src/azure-cli/azure/cli/command_modules/acs/tests/latest/test_aks_commands.py:2725]: test passed

File: src/azure-cli/azure/cli/command_modules/acs/tests/latest/test_aks_commands.py (hunk 4 above)
Comment (FumingZhang) [reply src/azure-cli/azure/cli/command_modules/acs/tests/latest/test_aks_commands.py:595]: test passed

**Review by yanzhudd — DISMISSED** [review]

File: src/azure-cli/azure/cli/command_modules/acs/custom.py [hunk 6]
```diff
@@ -585,12 +585,12 @@ def aks_create(
...
     nodepool_taints=None,
     node_osdisk_type=None,
-    node_osdisk_size=0,
+    node_osdisk_size=None,
     vm_set_type=None,
     zones=None,
     ppg=None,
     http_proxy_config=None,
-    max_pods=0,
+    max_pods=None,
```
[the remark this answers is not in the document]
Comment (FumingZhang) [reply src/azure-cli/azure/cli/command_modules/acs/custom.py:593]: This is a good question and I think this shouldn't be a breaking change as there's no behavioral change from a UX perspective.

The result of the following behavior is exactly the same as before and UT helps ensure this.
- If user does not set option `--max-pods` or set the value to `0`, the value of the corresponding filed in the request body is `null`.
- If user does not set option `--node-osdisk-size`, the value of the corresponding filed in the request body is `null`.
- If user set option `--node-osdisk-size 0`, the value of the corresponding filed in the request body is `0`.

The substantive result of this change is to remove the misleading annotation of the default value of these 2 options in the document.

image
image
https://learn.microsoft.com/en-us/cli/azure/aks?view=azure-cli-latest#az-aks-create

[the remark this answers is not in the document]
Comment (zhoxing-ms) [reply src/azure-cli/azure/cli/command_modules/acs/custom.py:593]: > If user does not set option --node-osdisk-size, the value of the corresponding filed in the request body is null.
If user set option --node-osdisk-size 0, the value of the corresponding filed in the request body is 0.

@FumingZhang I would like to confirm if values `null` or `0` have different impacts on the customer's business?

[the remark this answers is not in the document]
Comment (FumingZhang) [reply src/azure-cli/azure/cli/command_modules/acs/custom.py:593]: - For `--max-pods`
  - `null` is acceptable by the server and server will backfill the default value according to the network plugin used by the cluster.
  - `0` is invalid, server will return 400
- For `--node-osdisk-size`
  - Both `null` and `0` are acceptable by the server and server will backfill the default value (to 128).

Once again, I could confirm that after this change, the interaction between cli and server will not change when the user input remains unchanged.

[the remark this answers is not in the document]
Comment (zhoxing-ms) [reply src/azure-cli/azure/cli/command_modules/acs/custom.py:593]: Got it, thanks~

**Review by zhoxing-ms — APPROVED** [review]

**Pull request merged**